packages/frontend-api/src/Model/Mutation/Order/CreateOrderMutation.php

```diff
@@ -0,0 +1,116 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Shopsys\FrontendApiBundle\Model\Mutation\Order;
+
+use GraphQL\Error\UserError;
+use Overblog\GraphQLBundle\Definition\Argument;
+use Overblog\GraphQLBundle\Definition\Resolver\AliasedInterface;
+use Overblog\GraphQLBundle\Definition\Resolver\MutationInterface;
+use Overblog\GraphQLBundle\Validator\InputValidator;
+use Shopsys\FrameworkBundle\Model\Order\Mail\OrderMailFacade;
+use Shopsys\FrameworkBundle\Model\Order\Order;
+use Shopsys\FrontendApiBundle\Model\Order\OrderDataFacade;
+use Shopsys\FrontendApiBundle\Model\Order\PlaceOrderFacade;
+
+class CreateOrderMutation implements MutationInterface, AliasedInterface
+{
+    public const VALIDATION_GROUP_DIFFERENT_DELIVERY_ADDRESS = 'differentDeliveryAddress';
+    public const VALIDATION_GROUP_ON_COMPANY_BEHALF = 'onCompanyBehalf';
+
+    /**
+     * @var \Shopsys\FrontendApiBundle\Model\Order\OrderDataFacade
+     */
+    protected $orderDataFacade;
+
+    /**
+     * @var \Shopsys\FrontendApiBundle\Model\Order\PlaceOrderFacade
+     */
+    protected $placeOrderFacade;
+
+    /**
+     * @var \Shopsys\FrameworkBundle\Model\Order\Mail\OrderMailFacade
+     */
+    protected $orderMailFacade;
+
+    /**
+     * @param \Shopsys\FrontendApiBundle\Model\Order\OrderDataFacade $orderDataFacade
+     * @param \Shopsys\FrontendApiBundle\Model\Order\PlaceOrderFacade $placeOrderFacade
+     * @param \Shopsys\FrameworkBundle\Model\Order\Mail\OrderMailFacade $orderMailFacade
+     */
+    public function __construct(
+        OrderDataFacade $orderDataFacade,
+        PlaceOrderFacade $placeOrderFacade,
+        OrderMailFacade $orderMailFacade
+    ) {
+        $this->orderDataFacade = $orderDataFacade;
+        $this->placeOrderFacade = $placeOrderFacade;
+        $this->orderMailFacade = $orderMailFacade;
+    }
+
+    /**
+     * @param \Overblog\GraphQLBundle\Definition\Argument $argument
+     * @param \Overblog\GraphQLBundle\Validator\InputValidator $validator
+     * @return \Shopsys\FrameworkBundle\Model\Order\Order
+     */
+    public function createOrder(Argument $argument, InputValidator $validator): Order
+    {
+        $validationGroups = $this->computeValidationGroups($argument);
+        $validator->validate($validationGroups);
+
+        $orderData = $this->orderDataFacade->createOrderDataFromArgument($argument);
+        $quantifiedProducts = $this->orderDataFacade->createQuantifiedProductsFromArgument($argument);
+
+        $order = $this->placeOrderFacade->placeOrder($orderData, $quantifiedProducts);
+
+        try {
+            $this->sendEmail($order);
+        } catch (\Shopsys\FrameworkBundle\Model\Mail\Exception\MailException $e) {
+            throw new UserError('Unable to send some emails, please contact us for order verification.');
```

Comment:
This is weird. I create an order (with a flush) and I can't send an email and I get a mistake not to send an email, but I won't get an order that is correct saved. I don't know if we should return order with error? :thinking: 

What is your opinion?

Reply:
I see... I tried some testing and the only way I was able to raise this error is when I set the message body to null.
If I stop SMTP container, and even when I disable email spooling, this error does not occur (which is strange to me btw.)

So I think it's more a failsafe and this should not occur.

Returning order and somehow errors seems reasonable, but I'm not sure how to do it.
Maybe something like https://blog.logrocket.com/handling-graphql-errors-like-a-champ-with-unions-and-interfaces/
but that would mean you need to add the "error part" into your query. I'm not sure if I want to do that.

The ultimate goal to me is to separate sending emails into some async queue and get rid of this problem once and for all.

Reply:
Ok, thanks you for testing and for your opinion. Btw I agree with async queue, like a Rabbit, AWS SQS or Google CTQ.